examples/tpm_io.c

```diff
@@ -623,6 +630,95 @@
         return ret;
     }
 
+#elif defined(__QNX__) || defined(__QNXNTO__)
+
+    /* customization to QNX SPI master to allow keeping 
+        CS asserted (low) till spi_close. See IDE/QNX for spi_master patch */
+    #ifndef SPI_MODE_MAN_CS
+    #define SPI_MODE_MAN_CS (1 << 17) /* Manual Chip select */
+    #endif
+    #ifndef SPI_MODE_CLEAR_CS
+    #define SPI_MODE_CLEAR_CS (1 << 18) /* Clear all chip selects (used with SPI_MODE_MAN_CS) */
+    #endif
+    static int TPM2_IoCb_QNX_SPI(TPM2_CTX* ctx, const byte* txBuf,
+        byte* rxBuf, word16 xferSz, void* userCtx)
+    {
+        int fd;
+        int ret = TPM_RC_FAILURE;
+        int status;
+    #ifdef WOLFTPM_CHECK_WAIT_STATE
+        int timeout = TPM_SPI_WAIT_RETRY;
+    #endif
+        spi_cfg_t cfg;
+
+        /* open device */
+        fd = spi_open(TPM2_SPI_DEV);
+        if (fd == -1) {
+            return TPM_RC_FAILURE;
+        }
+        XMEMSET(&cfg, 0, sizeof(cfg));
+        cfg.mode = 8; /* 8-bits - CPOL=0/CPHA=0 */
+        cfg.mode |= SPI_MODE_MAN_CS; /* manual chip select - leave asserted */
+        cfg.clock_rate = TPM2_SPI_HZ;
+        status = spi_setcfg(fd, SPI_DEV_DEFAULT, &cfg);
+        if (status != 0) {
+            spi_close(fd);
+            return ret;
+        }
+
+    #ifdef WOLFTPM_CHECK_WAIT_STATE
+        /* Send Header */
+        status = spi_xchange(fd, SPI_DEV_DEFAULT, 
+            (byte*)txBuf, rxBuf, TPM_TIS_HEADER_SZ);
+        if (status == -1) {
+            spi_close(fd);
```

Comment:
If I understand correctly, in the error cases, the CS could still be asserted. As part of clean up should the config be cleared as at the end on success?
```
        cfg.mode |= (SPI_MODE_MAN_CS | SPI_MODE_CLEAR_CS);
        (void)spi_setcfg(fd, SPI_DEV_DEFAULT, &amp;cfg);
```

Reply:
Correct. This manually clears the CS.

Reply:
If the call to `spi_xchange` fails here, I think it's possible that the CS is left asserted. Does `spi_close` guarantee deasserting or should we have a common clean block that is always executed (either by goto or falling through to a single `return`)?

Reply:
Fixed